oeps/oep-0045/decisions/0001-tutor-as-replacement-for-edx-configuration.rst

```diff
@@ -0,0 +1,328 @@
+
+Note that it is perfectly possible to bypass Tutor entirely when it comes to cluster management. One could use its image
+creation engine to build the requisite Docker images, then manage the Kubernetes cluster with something like Helm: there
+are reports of `this exact thing being done
+<https://discuss.openedx.org/t/lets-talk-about-the-native-installation/3269/2>`__).
+
+
+Criterion 2. Automated deployment and testing
+.............................................
+
+Launching a preconfigured instance from scratch is simple, as Tutor can either `load configuration
+<https://docs.tutor.overhang.io/configuration.html#configuration>`__ from environment variables, from locally stored
+YAML files, or via command line parameters. Automating a launch is therefore just as simple as launching an
+``edx/configuration`` Ansible run.
+
+Nevertheless, it is not currently possible to run edx-platform tests in a Tutor-created container, so automation for
+this particular purpose will require work.
+
+
+Criterion 3. Support for the latest open release
+................................................
+
+Tutor supports the latest Open edX release, Koa, out of the box.  Furthermore, for at least the past two releases,
+development has been done concomitantly with the actual Open edX release itself.
+
+There's also built-in support for upgrading single servers, either `automatically
+<https://docs.tutor.overhang.io/local.html?highlight=upgrade#upgrading-from-earlier-versions>`__ or `manually
+<https://github.com/overhangio/tutor/blob/master/tutor/commands/local.py#L91>`__.  This, however, `does not apply to
+Kubernetes support <https://github.com/overhangio/tutor/blob/master/tutor/commands/k8s.py#L281-L288>`__.  While the
+corresponding single-server upgrade procedure can be used as an authoritative reference for how to achieve it manually
+on multi-server deployments, ideally there would be specific documentation on it.
+
+It is important to note that Tutor does not share any code with ``edx/configuration``, nor was it built to update
+containers in place with tools such as Ansible. As such, like other modern container deployment solutions, there is no
+way to update code other than first updating the container image then rolling it out.
+
+
+Criterion 4. Support for master
+...............................
+
+Regis Behmo `reports <https://discuss.openedx.org/t/lets-talk-about-the-native-installation/3269/5>`__ that some work in
+this direction has already `been done <https://github.com/overhangio/tutor/tree/edge>`__, but at the time of writing it
+is a couple of months behind the latest version of Tutor.
+
+This is clearly an area where contributions are needed. It will likely require one or more volunteers to help maintain
+master and fix breakages, as those would inevitably happen. Continuous deployment and some level of integration testing
+would also be required, as the volunteer(s) would need to be warned as soon as possible about failures.
```

Comment:
As I mentioned elsewhere, Tutor already has a working and performant gitlab-based CI. Currently, the CI workers run on Kubernetes, which means that they are scalable. The only issue I see is the cost of CI. It will be expensive (500-1000€/month) to run a full test suite for every PR. This is a place where Open edX hosting providers can contribute -- either financially, or by running CI workers on their infrastructure. In the latter case, it would make sense to display a "sponsored by" message at the top of every CI run.

In other words, having a CI running on the master branch is both desirable, feasible, and a good opportunity for collaboration.

Reply:
@regisb That sounds like a great idea. If we do the switch to Tutor, I would definitely be happy to get OpenCraft to contribute on this.

Reply:
I would also be happy to get eduNEXT to collaborate on this once we reach that stage,

Reply:
@regisb:

> The only issue I see is the cost of CI. It will be expensive (500-1000€/month) to run a full test suite for every PR. This is a place where Open edX hosting providers can contribute -- either financially, or by running CI workers on their infrastructure.

At [City Cloud](https://citycloud.com/) we'd be able to contribute what I think would be the required Kubernetes capacity for GitLab runners. We already contribute CI infrastructure for Open Infrastructure Foundation projects, so this shouldn't be very tough to argue internally, plus my team and I are already quite familiar with operating gitlab-runner instances on Kubernetes (or, for that matter, on dedicated VMs — we use both).

Reply:
@regisb As we're nearing agreement on this OEP I'd like to come back to this. What's the best way to move this forward? Can I suggest you send me an email (at your convenience — no rush at all) so we can discuss how to best make this happen?

Reply:
Thanks again for oferring your help @fghaas! At this point the following items would need to be prioritized:

1. Creating a "nightly" Tutor branch that would run the master branches of Open edX.
2. Getting the unit tests to run this nightly branch.
3. Bonus: Getting the integration tests to run on this nightly branch.
4. Create a CI that will actually run the tests.

Items 1 and 2 are obviously the most difficult steps here. I'm currently talking with edX to see how we can prioritize this together.

Reply:
Oh I understand that, but maybe we could start helping out by providing additional GitLab runners for the existing CI pipeline?